### PR TITLE
Replaced DateTime with IGameTiming for lobby timing

### PR DESCRIPTION
**Pull request by DTanxxx**
Replaced all the DateTime.UtcNow with IoCManager.Resolve<IGameTiming>().RealTime

### CODE REVIEW

**Review by Zumorica — CHANGES_REQUESTED** [review]
You don't seem to understand what `IGameTiming` is. You can't just replace all `DateTime` usages for literally `IGameTiming` and call it a day... `IGameTiming` provides you with ways to get the current time. What you're doing is copying the reference to that object around, and treating them as if they were "snapshots" of the class in the moment you copied them. This is simply not the case. That's not how any of this works. You must use the `RealTime` property, which returns a `TimeSpan`, in `IGameTiming` to get the time at the moment you called the property. 

In short, you should have replaced `DateTime` by `TimeSpan` with the help of `IGameTiming`, not literally replace all `DateTime` usages by `IGameTiming`...

**Comment by DTanxxx** [comment]
It looked pretty wrong to me as well, thanks for the clarification. I submitted another commit before seeing your comment so I will need to make some more adjustments later. I have some questions though:
- for the `StartTime` property in classes such as ClientGameTicker, should I replace its type with `TimeSpan` rather than `DateTime`? 
- for net messages such as `MsgTickerLobbyCountdown`, should I also change its `StartTime` to a `TimeSpan`? Does the message need data on the exact date and time it was sent or does it only require the amount of time since the simulation started?
Thanks.

File: Content.Client/State/LobbyState.cs [hunk 1]
```diff
@@ -133,7 +135,7 @@ public override void FrameUpdate(FrameEventArgs e)
             }
             else
             {
-                var difference = _clientGameTicker.StartTime - DateTime.UtcNow;
+                var difference = _gameTiming.CurTime;
```
Comment (Zumorica) [inline Content.Client/State/LobbyState.cs:138]: This is incorrect. This was supposed to be a difference of time between `StartTime` and the current time. Now, you set it to just the current time.

File: Content.Server/GameTicking/GameTicker.cs [hunk 2]
```diff
@@ -140,6 +141,8 @@ public override void Initialize()
 
             SetStartPreset(_configurationManager.GetCVar(CCVars.GameLobbyDefaultPreset));
 
+            _roundInitializedTimeUtc = DateTime.UtcNow;
```
Comment (Zumorica) [inline Content.Server/GameTicking/GameTicker.cs:144]: When you've changed `_roundInitializedTimeUtc` to use `TimeSpan`, this should be set to `_gameTiming.RealTime`.

File: Content.Server/GameTicking/GameTicker.cs [hunk 3]
```diff
@@ -166,7 +169,7 @@ public void Update(FrameEventArgs frameEventArgs)
 
             if (RunLevel != GameRunLevel.PreRoundLobby ||
                 Paused ||
-                _roundStartTimeUtc > DateTime.UtcNow ||
+                _roundStartTimeTicks > IoCManager.Resolve<IGameTiming>().RealTime.Ticks ||
```
Comment (Zumorica) [inline Content.Server/GameTicking/GameTicker.cs:172]: First of all, you already have a dependency for `IGameTiming`, so use that instead of resolving here every tick...
Second of all, you should just compare `_roundInitializedTimeUtc` to `_gameTiming.RealTime` instead of using these "ticks".

File: Content.Server/GameTicking/GameTicker.cs [hunk 4]
```diff
@@ -203,7 +206,7 @@ public void RestartRound()
                 if (PlayerManager.PlayerCount == 0)
                     _roundStartCountdownHasNotStartedYetDueToNoPlayers = true;
                 else
-                    _roundStartTimeUtc = DateTime.UtcNow + LobbyDuration;
+                    _roundStartTimeTicks = IoCManager.Resolve<IGameTiming>().RealTime.Ticks + LobbyDuration.Ticks;
```
Comment (Zumorica) [inline Content.Server/GameTicking/GameTicker.cs:209]: Again, you have an `IGameTiming` dependency so use that instead of resolving every time. And don't use ticks please.

File: Content.Server/GameTicking/GameTicker.cs [hunk 5]
```diff
@@ -508,10 +511,10 @@ public bool DelayStart(TimeSpan time)
                 return false;
             }
 
-            _roundStartTimeUtc += time;
+            _roundStartTimeTicks += time.Ticks;
 
             var lobbyCountdownMessage = _netManager.CreateNetMessage<MsgTickerLobbyCountdown>();
-            lobbyCountdownMessage.StartTime = _roundStartTimeUtc;
+            lobbyCountdownMessage.StartTime = new DateTime(_roundStartTimeTicks + _roundInitializedTimeUtc.Ticks, DateTimeKind.Utc);
```
Comment (Zumorica) [inline Content.Server/GameTicking/GameTicker.cs:517]: At this point, just change that to use `TimeSpan`...

File: Content.Server/GameTicking/GameTicker.cs [hunk 6]
```diff
@@ -531,15 +534,15 @@ public bool PauseStart(bool pause = true)
 
             if (pause)
             {
-                _pauseTime = DateTime.UtcNow;
+                _pauseTime = IoCManager.Resolve<IGameTiming>();
```
Comment (Zumorica) [inline Content.Server/GameTicking/GameTicker.cs:537]: No. Use the dependency you have, and get `RealTime` instead of setting `_pauseTime` to `IGameTiming`.

File: Content.Server/GameTicking/GameTicker.cs [hunk 7]
```diff
@@ -531,15 +534,15 @@ public bool PauseStart(bool pause = true)
...
             }
             else if (_pauseTime != default)
             {
-                _roundStartTimeUtc += DateTime.UtcNow - _pauseTime;
+                _roundStartTimeTicks += IoCManager.Resolve<IGameTiming>().RealTime.Ticks - _pauseTime.RealTime.Ticks;
```
Comment (Zumorica) [inline Content.Server/GameTicking/GameTicker.cs:541]: You have a dependency for this. Use it. Also don't use ticks.

File: Content.Server/GameTicking/GameTicker.cs [hunk 8]
```diff
@@ -531,15 +534,15 @@ public bool PauseStart(bool pause = true)
...
             }
 
             var lobbyCountdownMessage = _netManager.CreateNetMessage<MsgTickerLobbyCountdown>();
-            lobbyCountdownMessage.StartTime = _roundStartTimeUtc;
+            lobbyCountdownMessage.StartTime = new DateTime(_roundStartTimeTicks + _roundInitializedTimeUtc.Ticks, DateTimeKind.Utc);
```
Comment (Zumorica) [inline Content.Server/GameTicking/GameTicker.cs:545]: I've already said this before, don't use `DateTime`. Change that message to use `TimeSpan`.

File: Content.Server/GameTicking/GameTicker.cs [hunk 9]
```diff
@@ -731,7 +734,7 @@ protected override void PlayerStatusChanged(object sender, SessionStatusEventArg
                     if (LobbyEnabled && _roundStartCountdownHasNotStartedYetDueToNoPlayers)
                     {
                         _roundStartCountdownHasNotStartedYetDueToNoPlayers = false;
-                        _roundStartTimeUtc = DateTime.UtcNow + LobbyDuration;
+                        _roundStartTimeTicks = IoCManager.Resolve<IGameTiming>().RealTime.Ticks + LobbyDuration.Ticks;
```
Comment (Zumorica) [inline Content.Server/GameTicking/GameTicker.cs:737]: [same thing as before]

File: Content.Server/GameTicking/GameTicker.cs [hunk 10]
```diff
@@ -977,7 +980,7 @@ private MsgTickerLobbyStatus _getStatusMsg(IPlayerSession session)
             _playersInLobby.TryGetValue(session, out var status);
             var msg = _netManager.CreateNetMessage<MsgTickerLobbyStatus>();
             msg.IsRoundStarted = RunLevel != GameRunLevel.PreRoundLobby;
-            msg.StartTime = _roundStartTimeUtc;
+            msg.StartTime = new DateTime(_roundStartTimeTicks + _roundInitializedTimeUtc.Ticks, DateTimeKind.Utc);
```
Comment (Zumorica) [inline Content.Server/GameTicking/GameTicker.cs:983]: [same thing as before]

**Review by Zumorica — CHANGES_REQUESTED** [review]

**Comment by DTanxxx** [comment]
Hi Zumorica,

Should I also replace DateTime with TimeSpan in files related to PlayerSession (in RobustToolbox module)?

**Comment by Zumorica** [comment]
No, that one is fine since it isn't used for a countdown or anything of sorts.

**Comment by DTanxxx** [comment]
Got it, thanks.

**Comment by AJCM-git** [comment]
Will this get finished? will close in a week if no response

**Comment by metalgearsloth** [comment]
PJB did this.
